From dac4b0bf946e2b1189ab6fc15d5ae59cc1c67150 Mon Sep 17 00:00:00 2001
From: Ivan Chiang <chiangi@google.com>
Date: Tue, 16 Nov 2021 08:56:59 +0000
Subject: [PATCH] Avoid apps to get global access with table uri

If the app is granted the read UriPerimssion on the uri without id,
don't allow the app to get global access. Then, the app gets the same
query result of the uri without read UriPermission.

Test: atest --test-mapping packages/providers/MediaProvider
Bug: 197302116
Change-Id: Id6349b7b49cf682c6703333bc96f1b7046b7700e
Merged-In: Id6349b7b49cf682c6703333bc96f1b7046b7700e
(cherry picked from commit a728dfa19c2601b576a718fd9f79916bccd07c0e)
Merged-In:Id6349b7b49cf682c6703333bc96f1b7046b7700e
---
 .../providers/media/MediaProvider.java        | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 626bbb8e..40a4e258 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -7050,6 +7050,26 @@ public class MediaProvider extends ContentProvider {
                 if (mCallingIdentity.get().isOwned(id)) {
                     return true;
                 }
+                break;
+            default:
+                // continue below
+        }
+
+        // Check whether the uri is a specific table or not. Don't allow the global access to these
+        // table uris
+        switch (table) {
+            case AUDIO_MEDIA:
+            case IMAGES_MEDIA:
+            case VIDEO_MEDIA:
+            case DOWNLOADS:
+            case FILES:
+            case AUDIO_ALBUMS:
+            case AUDIO_ARTISTS:
+            case AUDIO_GENRES:
+            case AUDIO_PLAYLISTS:
+                return false;
+            default:
+                // continue below
         }
 
         // Outstanding grant means they get access
-- 
2.17.1

